pkg/virt-operator/resource/apply/apps.go

```diff
@@ -87,7 +109,148 @@ func (r *Reconciler) syncDeployment(origDeployment *appsv1.Deployment) (*appsv1.
 	return deployment, nil
 }
 
-func (r *Reconciler) syncDaemonSet(daemonSet *appsv1.DaemonSet) error {
+func setMaxUnavailable(daemonSet *appsv1.DaemonSet, maxUnavailable intstr.IntOrString) {
+	daemonSet.Spec.UpdateStrategy.RollingUpdate = &appsv1.RollingUpdateDaemonSet{
+		MaxUnavailable: &maxUnavailable,
+	}
+}
+
+func generateDaemonSetPatch(oldDs, newDs *appsv1.DaemonSet) ([]byte, error) {
+	newSpec, err := json.Marshal(newDs.Spec)
+	if err != nil {
+		return nil, err
+	}
+
+	ops, err := getPatchWithObjectMetaAndSpec([]string{
+		fmt.Sprintf(testGenerationJSONPatchTemplate, oldDs.ObjectMeta.Generation),
+	}, &newDs.ObjectMeta, newSpec)
+	if err != nil {
+		return nil, err
+	}
+	return generatePatchBytes(ops), nil
+}
+
+func (r *Reconciler) patchDaemonSet(oldDs, newDs *appsv1.DaemonSet) (*appsv1.DaemonSet, error) {
+	patch, err := generateDaemonSetPatch(oldDs, newDs)
+	if err != nil {
+		return nil, err
+	}
+
+	newDs, err = r.clientset.AppsV1().DaemonSets(r.kv.Namespace).Patch(
+		context.Background(),
+		newDs.Name,
+		types.JSONPatchType,
+		patch,
+		metav1.PatchOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to update daemonset %+v: %v", oldDs, err)
+	}
+	return newDs, nil
+}
+
+func (r *Reconciler) getCanaryPod(daemonSet *appsv1.DaemonSet) *corev1.Pod {
+	for _, obj := range r.stores.InfrastructurePodCache.List() {
+		pod := obj.(*corev1.Pod)
+		owner := metav1.GetControllerOf(pod)
+
+		if owner != nil && owner.Name == daemonSet.Name && util.PodIsUpToDate(pod, r.kv) {
+			return pod
+		}
+	}
+	return nil
+}
+
+func (r *Reconciler) howManyUpdatedAndReadyPods(daemonSet *appsv1.DaemonSet) int32 {
+	var updatedReadyPods int32
+
+	for _, obj := range r.stores.InfrastructurePodCache.List() {
+		pod := obj.(*corev1.Pod)
+		owner := metav1.GetControllerOf(pod)
+
+		if owner != nil && owner.Name == daemonSet.Name && util.PodIsUpToDate(pod, r.kv) && util.PodIsReady(pod) {
+			updatedReadyPods++
+		}
+	}
+	return updatedReadyPods
+}
+
+func daemonHasDefaultRolloutStrategy(daemonSet *appsv1.DaemonSet) bool {
+	return getMaxUnavailable(daemonSet) == daemonSetDefaultMaxUnavailable.IntValue()
+}
+
+func (r *Reconciler) processCanaryUpgrade(cachedDaemonSet, newDS *appsv1.DaemonSet, forceUpdate bool) (bool, error, CanaryUpgradeStatus) {
+	var updatedAndReadyPods int32
+	var status CanaryUpgradeStatus
+	done := false
+
+	isDaemonSetUpdated := util.DaemonSetIsUpToDate(r.kv, cachedDaemonSet) && !forceUpdate
+	desiredReadyPods := cachedDaemonSet.Status.DesiredNumberScheduled
+	if isDaemonSetUpdated {
+		updatedAndReadyPods = r.howManyUpdatedAndReadyPods(cachedDaemonSet)
+	}
+
+	switch {
+	case updatedAndReadyPods == 0:
+		if !isDaemonSetUpdated {
+			// start canary upgrade
+			setMaxUnavailable(newDS, intstr.FromInt(1))
```

Comment:
for consistency, should this be `setMaxUnavailable(newDS, daemonSetDefaultMaxUnavailable)` instead of 1?

Reply:
Will change it in the next push, thanks for spotting this!